src/java.net.http/share/classes/jdk/internal/net/http/common/SSLTube.java

```diff
@@ -619,43 +619,74 @@ final class SSLSubscriptionWrapper implements Flow.Subscription {
         private volatile boolean cancelled;
 
         void setSubscription(Flow.Subscription sub) {
-            long demand = writeDemand.get(); // FIXME: isn't it a racy way of passing the demand?
-            delegate = sub;
-            if (debug.on())
-                debug.log("setSubscription: demand=%d, cancelled:%s", demand, cancelled);
+            long demand;
+            // Avoid race condition and requesting demand twice if
+            // request() runs concurrently with setSubscription()
+            boolean cancelled;
+            synchronized (this) {
+                demand = writeDemand.get();
+                delegate = sub;
+                cancelled = this.cancelled;
+            }
+            if (debug.on()) {
+                debug.log("setSubscription: demand=%d, cancelled:%s, new subscription %s",
+                        demand, cancelled, sub);
+            }
 
             if (cancelled)
-                delegate.cancel();
+                sub.cancel();
             else if (demand > 0)
                 sub.request(demand);
         }
 
         @Override
         public void request(long n) {
-            writeDemand.increase(n);
-            if (debug.on()) debug.log("request: n=%d", n);
-            Flow.Subscription sub = delegate;
-            if (sub != null && n > 0) {
-                sub.request(n);
+            final long demand = n;
```

Comment:
Hello Daniel, is this `demand` variable necessary? From what I see, we could just keep using `n` wherever we use `demand` in this method. I think that would also avoid the confusion between `demanded` and this `demand` variables that are used in this method.

Reply:
OK